aesara/tensor/nlinalg.py

```diff
@@ -823,6 +824,87 @@ def tensorsolve(a, b, axes=None):
     return TensorSolve(axes)(a, b)
 
 
+class MatMul(Op):
+    __props__ = ("dtype",)
+
+    def __init__(self, dtype=None):
+        self.dtype = dtype
+
+    def make_node(self, a, b):
+        a = at.as_tensor_variable(a)
+        b = at.as_tensor_variable(b)
+
+        if 0 in (a.ndim, b.ndim):
+            raise ValueError("inputs to `matmul` cannot be scalar.")
+        elif a.ndim >= 2 and b.ndim >= 2:
+            out = at.TensorType(
+                dtype=self.dtype,
+                broadcastable=(False,) * max(a.ndim, b.ndim),
+            )()
+        else:
+            out_ndim = b.ndim - 1 if a.ndim == 1 else a.ndim - 1
+            out = at.TensorType(
+                dtype=self.dtype,
+                broadcastable=(False,) * out_ndim,
+            )()
+        return Apply(self, [a, b], [out])
+
+    def perform(self, node, inputs, outputs):
+        x1, x2 = inputs
+        outputs[0][0] = np.matmul(x1, x2, dtype=self.dtype)
+
+    def infer_shape(self, fgraph, node, shapes):
+        x1_shape, x2_shape = shapes
+        x1, x2 = node.inputs
+
+        if x1.ndim == 1 and x2.ndim == 1:
+            return [()]
+        elif x1.ndim == 1:
+            return [x2_shape[:-2] + x2_shape[-1:]]
+        elif x2.ndim == 1:
+            return [x1_shape[:-1]]
+        elif x1.ndim == 2 and x2.ndim == 2:
+            return [x1_shape[:-1] + x2_shape[-1:]]
+        elif x2.ndim == 2:
+            return [x1_shape[:-2] + x1_shape[-2:-1] + x2_shape[-1:]]
+        elif x1.ndim == 2:
+            return [x2_shape[:-2] + x1_shape[-2:-1] + x2_shape[-1:]]
+        else:
+            return [
+                broadcast_shape(x1[..., 0, 0], x2[..., 0, 0])
```

Comment:
It might work, but this explicit case handling will definitely produce simpler graphs in most cases.  Hopefully, we'll come up with an efficient solution to canonicalizing the results of `Op.infer_shape`, but, at this point, we don't have one implemented, so local "optimizations" like this aren't unwelcome.

In summary, if the output can be represented that succinctly it really should be, but due to other limitations of this library, it's currently acceptable.

Reply:
```  File "/home/zoj/.conda/envs/aesara/lib/python3.7/site-packages/numpy/lib/stride_tricks.py", line 470, in broadcast_shapes
    arrays = [np.empty(x, dtype=[]) for x in args]
  File "/home/zoj/.conda/envs/aesara/lib/python3.7/site-packages/numpy/lib/stride_tricks.py", line 470, in <listcomp>
    arrays = [np.empty(x, dtype=[]) for x in args]
TypeError: 'TensorVariable' object cannot be interpreted as an integer
```
numpy is not too happy about receiving TensorVariable in the np.broadcast_shapes implementation.  Here is a reproduction example:
```python
In [21]: np.empty((at.as_tensor(3), at.as_tensor(4)), dtype=[])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-a80110ee1c26> in <module>
----> 1 np.empty((at.as_tensor(3), at.as_tensor(4)), dtype=[])

TypeError: 'TensorConstant' object cannot be interpreted as an integer
```

Reply:
but using the aesara implementation produces its own errors:
```
aesara/tensor/basic_opt.py:1008: KeyError
----------------------------------------------------------------- Captured log call ------------------------------------------------------------------
ERROR    aesara.graph.opt:opt.py:221 SeqOptimizer apply <aesara.tensor.basic_opt.ShapeOptimizer object at 0x7f21fa584310>
ERROR    aesara.graph.opt:opt.py:222 Traceback:
ERROR    aesara.graph.opt:opt.py:223 Traceback (most recent call last):
  File "/home/zoj/dev/aesara/aesara/graph/opt.py", line 279, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "/home/zoj/dev/aesara/aesara/graph/opt.py", line 101, in optimize
    self.add_requirements(fgraph)
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1411, in add_requirements
    fgraph.attach_feature(ShapeFeature())
  File "/home/zoj/dev/aesara/aesara/graph/fg.py", line 569, in attach_feature
    attach(self)
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1221, in on_attach
    self.on_import(fgraph, node, reason="on_attach")
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1288, in on_import
    self.set_shape(r, s)
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1053, in set_shape
    shape_vars.append(self.unpack(s[i], r))
  File "/home/zoj/dev/aesara/aesara/tensor/basic_opt.py", line 1008, in unpack
    s_i = self.shape_of[x][i]
KeyError: Subtensor{::, ::, int64, int64}.0
```
@brandonwillard do you have an idea of what might causing this?

Reply:
> numpy is not too happy about receiving TensorVariable in the np.broadcast_shapes implementation. Here is a reproduction example:

You definitely don't want to pass NumPy Aesara variables.  Use the equivalent Aesara function instead.

Reply:
That doesn't work either as shown in the stack trace right above your post. Do you know why its throwing an error like that?

Reply:
the shape inference works if I use something like `(3, 2, 4, 5), (3, 2, 5, 7)`, but fails for `(3, 2, 4, 5), (2, 5, 7)`

Reply:
> but using the aesara implementation produces its own errors:

The error is saying that `ShapeFeature`, which keeps track of all the shapes in a `FunctionGraph`, is missing an entry for a `Variable` it's encountered.  This can happen when `ShapeFeature` comes across a `Variable` that wasn't properly added to the `FunctionGraph`, because such variables wouldn't have triggered `FunctionGraph`'s callbacks and&mdash;as a result&mdash;wouldn't have been picked up by `ShapeFeature`.

If this is caused by the use of `broadcast_shape` in `MatMul.infer_shape`, then it's probably because `broadcast_shape` is introducing those elements.  If I recall, `broadcast_shape` will use the shapes of its arguments, and any time a shape is requested for a `Variable` (e.g. 
 via the `TensorVariable.shape` method) a `Shape` `Op` is created in the resulting graph.  The function of `ShapeFeature` is to replace those `Shape`s with the results of the relevant `Op`'s `Op.infer_shape` method (like the one you're writing).

What's likely happening is that your `MatMul.infer_shape` is creating a graph that computes the shape of a `Variable`, and that graph itself contains `Shape` `Op`s that need to be converted by `ShapeFeature`, but the inputs to some of those `Shape` `Op`s are new `Variable`s that were just created inside of `MatMul.infer_shape`, so `ShapeFeature` has never seen them and raises that `KeyError`.

The solution is to always avoid using `Shape` `Op`s inside of the graphs produced by an `Op.infer_shape`.  This is part of the reason why `Op.infer_shape` has an argument that specifies the shapes of the `Op`s inputs, because those are all one should need in order to compute the shape of the `Op`'s outputs.

Unfortunately, this restriction and approach are largely due to some very old design issues.  I actually opened an issue earlier today about this very situation: #749.

Reply:
Never mind. I was completely unaware that `broadcast_shape` has a `arrays_are_shapes` parameter. everything works now. Embarrassing moment there *face palm*.

Reply:
It's nothing to be embarrassed about; the source of the confusion is primarily due to an unnecessarily confusing and convoluted shape handling system.

Reply:
Yeah definitely nothing to be embarrassed about, the API is also very confusing here.